# fullscreen and default size follow the Xinerama screen under the cursor
--- a/configure.ac
+++ b/configure.ac
@@ -35,6 +35,8 @@
 AC_SEARCH_LIBS(socket, socket)
 AC_SEARCH_LIBS(inet_aton, resolv)
 
+AC_CHECK_LIB(Xinerama, XineramaQueryScreens, AC_DEFINE(HAVE_XINERAMA) LIBS="$LIBS -lXinerama", [], [])
+
 AC_CHECK_HEADER(sys/select.h, AC_DEFINE(HAVE_SYS_SELECT_H))
 AC_CHECK_HEADER(sys/modem.h, AC_DEFINE(HAVE_SYS_MODEM_H))
 AC_CHECK_HEADER(sys/filio.h, AC_DEFINE(HAVE_SYS_FILIO_H))
--- a/xwin.c
+++ b/xwin.c
@@ -37,6 +37,9 @@
 #ifdef HAVE_XRANDR
 #include <X11/extensions/Xrandr.h>
 #endif
+#ifdef HAVE_XINERAMA
+#include <X11/extensions/Xinerama.h>
+#endif
 
 #ifdef __APPLE__
 #include <sys/param.h>
@@ -872,6 +875,67 @@
 #define LOUT24(o, x) { *(o++) = x; *(o++) = x >> 8; *(o++) = x >> 16; }
 #define LOUT32(o, x) { *(o++) = x; *(o++) = x >> 8; *(o++) = x >> 16; *(o++) = x >> 24; }
 
+#ifdef HAVE_XINERAMA
+void
+ScreenSize(XRectangle *screen)
+{
+	int screens;
+	int event_base;
+	int error_base;
+	int x, y, i, __unused_int;
+	unsigned int __unused_uint;
+	Window __unused_win;
+
+	if ((XineramaQueryExtension(g_display, &event_base, &error_base)) && (XineramaIsActive(g_display)))
+	{
+		XineramaScreenInfo *screeninfo;
+
+		/* getting current cursor position */
+		XQueryPointer(g_display, RootWindowOfScreen(g_screen), &__unused_win, &__unused_win, &x, &y, &__unused_int, &__unused_int, &__unused_uint);
+
+		/* Get the Xinerama screen infomation */
+		screeninfo = XineramaQueryScreens(g_display, &screens);
+
+		/* Search for the appropriate screen */
+		i = 0;
+		while (i < screens
+			&& !(screeninfo[i].x_org <= x
+				&& screeninfo[i].y_org <= y
+				&& screeninfo[i].x_org + screeninfo[i].width >= x
+				&& screeninfo[i].y_org + screeninfo[i].height >= y))
+			i++;
+		if (i >= screens)
+			i = 0;
+
+		/* Position according to the present screen */
+		screen->x = screeninfo[i].x_org;
+		screen->y = screeninfo[i].y_org;
+		screen->width = screeninfo[i].width;
+		screen->height = screeninfo[i].height;
+
+		/* Free allocated memory */
+		XFree(screeninfo);
+	}
+	else
+	{
+		/* Xinerama is not in use, default to the XLib screensize call. */
+		screen->x = 0;
+		screen->y = 0;
+		screen->width = WidthOfScreen(g_screen);
+		screen->height = HeightOfScreen(g_screen);
+	}
+}
+#else
+void
+ScreenSize(XRectangle *screen)
+{
+	screen->x = 0;
+	screen->y = 0;
+	screen->width = WidthOfScreen(g_screen);
+	screen->height = HeightOfScreen(g_screen);
+}
+#endif
+
 static uint32
 translate_colour(uint32 colour)
 {
@@ -2046,8 +2110,11 @@
 void
 ui_get_screen_size(uint32 * width, uint32 * height)
 {
-	*width = WidthOfScreen(g_screen);
-	*height = HeightOfScreen(g_screen);
+	XRectangle screen;
+
+	ScreenSize(&screen);
+	*width = screen.width;
+	*height = screen.height;
 }
 
 void
@@ -2209,10 +2276,20 @@
 	logger(GUI, Debug, "ui_create_window() width = %d, height = %d", width, height);
 
 	/* Handle -x-y portion of geometry string */
-	if (g_xpos < 0 || (g_xpos == 0 && (g_pos & 2)))
-		g_xpos = WidthOfScreen(g_screen) + g_xpos - width;
-	if (g_ypos < 0 || (g_ypos == 0 && (g_pos & 4)))
-		g_ypos = HeightOfScreen(g_screen) + g_ypos - height;
+	{
+		XRectangle screen;
+
+		ScreenSize(&screen);
+		if (g_fullscreen)
+		{
+			g_xpos = screen.x;
+			g_ypos = screen.y;
+		}
+		if (g_xpos < 0 || (g_xpos == 0 && (g_pos & 2)))
+			g_xpos = screen.width + g_xpos - width;
+		if (g_ypos < 0 || (g_ypos == 0 && (g_pos & 4)))
+			g_ypos = screen.height + g_ypos - height;
+	}
 
 	value_mask = get_window_attribs(&attribs);
 
@@ -2444,10 +2521,13 @@
 	if (g_fullscreen)
 	{
 		/* Since we need to create a fullscreen window we need to know screen size */
-		x = 0;
-		y = 0;
-		width = WidthOfScreen(g_screen);
-		height = HeightOfScreen(g_screen);
+		XRectangle screen;
+
+		ScreenSize(&screen);
+		x = screen.x;
+		y = screen.y;
+		width = screen.width;
+		height = screen.height;
 	}
 	else
 	{
@@ -3174,9 +3254,20 @@
 	/* Set up width and height for new session */
 	if (g_fullscreen || g_seamless_rdp)
 	{
-		/* follow root window size */
-		g_requested_session_width = WidthOfScreen(g_screen);
-		g_requested_session_height = HeightOfScreen(g_screen);
+		if (g_seamless_rdp)
+		{
+			/* seamless windows span the whole root window */
+			g_requested_session_width = WidthOfScreen(g_screen);
+			g_requested_session_height = HeightOfScreen(g_screen);
+		}
+		else
+		{
+			XRectangle screen;
+
+			ScreenSize(&screen);
+			g_requested_session_width = screen.width;
+			g_requested_session_height = screen.height;
+		}
 		if (g_window_size_type == PercentageOfScreen)
 		{
 			/* TODO: Implement percentage of screen */
